agent/consul/gateways/gateway_meta.go

```diff
@@ -0,0 +1,143 @@
+package gateways
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// gatewayMeta embeds both a BoundAPIGateway and its corresponding APIGateway.
+type gatewayMeta struct {
+	// Bound is the bound-api-gateway config entry for a given gateway.
+	Bound *structs.BoundAPIGatewayConfigEntry
+	// Gateway is the api-gateway config entry for the gateway.
+	Gateway *structs.APIGatewayConfigEntry
+}
+
+// getGatewayMeta queries the state store for an API Gateway and a Bound API
+// Gateway matching the given name and enterprise meta and returns a GatewayMeta
+// struct containing both.
+func getGatewayMeta(store *state.Store, name string, entMeta *acl.EnterpriseMeta) (*gatewayMeta, error) {
+	_, bound, err := store.ConfigEntry(nil, structs.BoundAPIGateway, name, entMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	_, gateway, err := store.ConfigEntry(nil, structs.APIGateway, name, entMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	return &gatewayMeta{
+		Bound:   bound.(*structs.BoundAPIGatewayConfigEntry),
```

Comment:
is it possible that `bound` is not actually a `*structs.BoundAPIGatewayConfigEntry` and this panics?

Reply:
I suppose it is possible. The `ConfigEntry` function is responsible for returning a value that should match the `kind` that was passed in.

Reply:
In this particular case, if it was something other than `*structs.BoundAPIGatewayConfigEntry` there would be a programming error deep within the bowels of our memdb storage layer, so I'm fine with the unchecked conversion here. It's pretty common in most of the read layers of different subsystems in Consul that are interacting with memdb, so if something were that wrong in our storage layer, arguably this should panic.

Reply:
gotcha, that's good to know